include/kernels/madwf_tensor.cuh

```diff
@@ -0,0 +1,205 @@
+#pragma once
+
+#include <color_spinor_field.h>
+#include <color_spinor_field_order.h>
+#include <shared_memory_cache_helper.cuh>
+#include <kernels/madwf_transfer.cuh>
+#include <block_reduce_helper.h>
+
+namespace quda
+{
+
+  namespace madwf_ml
+  {
+    template <class T> __device__ __host__ inline void block_reduce_x(T &f)
```

Comment:
What does `block_reduce_x` do that makes it distinct from `BlockReduce`?  If you are only wanting to do a reduction in the x thread dimension, we could likely modify the "batched" `BlockReduce` to meet the needs of this.  It is designed to do distinct reductions across the block z dimension and not the y dimension like is used here, but we could modify it accordingly easily enough.  The only place where the batched reductions are done is in the restrictor.

Reply:
Indeed we are just performing `BlockReduce` in the x direction.